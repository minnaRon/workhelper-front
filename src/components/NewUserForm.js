/**
 * Module, component
 * @description Form for the new user of the application to register.
 * Local states:
 * state {string} username - User's input for username
 * state {string} name - User's input for name
 * state {string} password - User's input for password
 * state {string} confirmPassword - User's input for confirm password
 * const {object} user - Current user, object fetched from the redux store
 * Form includes:
 * inputs:
 * username, length 4-50, required
 * name, length 3-50, required,
 * password, length 8-50, required
 * confirmPassword, required
 * buttons: submit, back
 * exports NewUserForm as default
 */
import { useState, useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useNavigate } from 'react-router-dom'
import { showNotification } from '../reducers/notificationRed'
import { addNewUser } from '../reducers/usersRed'

const NewUserForm = () => {
  const [username, setUsername] = useState('')
  const [name, setName] = useState('')
  const [password, setPassword] = useState('')
  const [confirmPassword, setConfirmPassword] = useState('')
  const user = useSelector((state) => state.user)
  const dispatch = useDispatch()
  const navigate = useNavigate()

  /**
   * @description Function handles new user's register inputs after the form is submitted.
   * @param {*} event - not in use here,
   * function uses state to store values: username, name, password and confirmPassword.
   * Checks if password is the same as confirmPassword, dispatch error notification if not and
   * checks if the length of the password is between 8-50, dispatch error notification if not.
   * When password is valid:
   * Dispatches register info to usersReducer's addNewUser function and
   * then resets local state values: username, name, password and confirmPassword.
   */
  const handleSubmit = async (event) => {
    event.preventDefault()
    if (password !== confirmPassword) {
      dispatch(showNotification('SALASANAT EIVÄT TÄSMÄÄ, TARKISTA SALASANAT', 'error'))
    } else if (password.length < 8 || password.length > 50) {
      dispatch(showNotification('SALASANAN PITUUS TULEE OLLA 8-50 MERKKIÄ', 'error'))
    } else {
      dispatch(addNewUser({ username, name, password }))
      if (user) {
        setUsername('')
        setName('')
        setPassword('')
        setConfirmPassword('')
      }
    }
  }

  /**
   * Hook useEffect
   * @description After the user is logged in navigates to workPlanView
   */
  useEffect(() => {
    if (user) {
      navigate('/todaysworkplan')
    }
  }, [dispatch])

  return (
    <div>
      <form onSubmit={handleSubmit}>
        <h2>UUSI KÄYTTÄJÄ</h2>
        <div>
          KÄYTTÄJÄNIMI
          <input
            type="text"
            id="newUserForm-username"
            value={username}
            minLength={4}
            maxLength={50}
            required
            onChange={({ target }) => setUsername(target.value)}
          />
        </div>
        <div>
          NIMI
          <input
            type="text"
            id="newUserForm-name"
            value={name}
            minLength={3}
            maxLength={50}
            required
            onChange={({ target }) => setName(target.value)}
          />
        </div>
        <div>
          SALASANA
          <input
            type="password"
            id="newUserForm-password"
            value={password}
            minLength={8}
            maxLength={50}
            required
            onChange={({ target }) => setPassword(target.value)}
          />
        </div>
        <div>
          SALASANA UUDELLEEN
          <input
            type="password"
            id="newUserForm-confirmPassword"
            value={confirmPassword}
            required
            onChange={({ target }) => setConfirmPassword(target.value)}
          />
        </div>
        <button type="submit" id="newUserForm-button-submit">
          TALLENNA JA KIRJAUDU
        </button>
      </form>
      <button id="newUserForm-button-back" onClick={() => navigate('/')}>
        TAKAISIN
      </button>
    </div>
  )
}

export default NewUserForm
