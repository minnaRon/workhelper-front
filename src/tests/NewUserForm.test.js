/**
 * snapshot
 */
import React from 'react'
import { Provider } from 'react-redux'
import { BrowserRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import '@testing-library/jest-dom/extend-expect'
import NewUserForm from '../components/NewUserForm'

const mockStore = configureStore([])

describe('<NewUserForm />', () => {
  let store
  let component

  beforeEach(() => {
    store = mockStore({
      users: [],
    })

    component = renderer.create(
      <Provider store={store}>
        <Router>
          <NewUserForm />
        </Router>
      </Provider>
    )
  })

  test('renders correct empty form (this fails also if local vocabulary changes', () => {
    expect(component.toJSON()).toMatchSnapshot()
  })
})
