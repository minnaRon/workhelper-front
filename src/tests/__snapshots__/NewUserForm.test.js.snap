// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewUserForm /> renders correct empty form (this fails also if local vocabulary changes 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <h2>
      UUSI KÄYTTÄJÄ
    </h2>
    <div>
      KÄYTTÄJÄNIMI
      <input
        id="newUserForm-username"
        maxLength={50}
        minLength={4}
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </div>
    <div>
      NIMI
      <input
        id="newUserForm-name"
        maxLength={50}
        minLength={3}
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </div>
    <div>
      SALASANA
      <input
        id="newUserForm-password"
        maxLength={50}
        minLength={8}
        onChange={[Function]}
        required={true}
        type="password"
        value=""
      />
    </div>
    <div>
      SALASANA UUDELLEEN
      <input
        id="newUserForm-confirmPassword"
        onChange={[Function]}
        required={true}
        type="password"
        value=""
      />
    </div>
    <button
      id="newUserForm-button-submit"
      type="submit"
    >
      TALLENNA JA KIRJAUDU
    </button>
  </form>
  <button
    id="newUserForm-button-back"
    onClick={[Function]}
  >
    TAKAISIN
  </button>
</div>
`;
